pkg/restic/repository_ensurer.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restic
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	arkv1api "github.com/heptio/ark/pkg/apis/ark/v1"
+	arkv1client "github.com/heptio/ark/pkg/generated/clientset/versioned/typed/ark/v1"
+	arkv1informers "github.com/heptio/ark/pkg/generated/informers/externalversions/ark/v1"
+	arkv1listers "github.com/heptio/ark/pkg/generated/listers/ark/v1"
+)
+
+// repositoryEnsurer ensures that Ark restic repositories are created and ready.
+type repositoryEnsurer struct {
+	repoLister arkv1listers.ResticRepositoryLister
+	repoClient arkv1client.ResticRepositoriesGetter
+
+	readyChansLock sync.Mutex
+	readyChans     map[string]chan *arkv1api.ResticRepository
+}
+
+func newRepositoryEnsurer(repoInformer arkv1informers.ResticRepositoryInformer, repoClient arkv1client.ResticRepositoriesGetter) *repositoryEnsurer {
```

Comment:
I would recommend changing the logic so we create a new one of these per `*restic.backupper`. Then we don't need a lock or multiple channels and can instead just have a single channel.

I'm even tempted to say we can do away with the informer+channel. Move the `initializeRepo` functionality from `*resticRepositoryController` directly into the `*restic.Backupper` or this `repositoryEnsurer`. Then it's a lot simpler to implement `EnsureRepo`:
1. Try to get the repo.
1. If it exists & is ready, done.
1. If it exists & isn't ready, wait (?) or call `Check()`. Return result of wait/check
1. If it doesn't exist, try to create it
1. If create succeeds, run the logic in `initializeRepo`

Reply:
If you want to avoid having multiple channels, we'd need to have one of these per call to `BackupPodVolumes` in the backupper, since a backup can have >1 namespace/repo. Not sure exactly what that would look like, since I don't think we'd want to be creating a new informer each time, but we can't be adding/removing event handlers while an existing one is running.

I'm not sure I like the idea of moving the repo init logic into the backupper or ensurer, but I think we need to talk it through in person.

Why don't you like what I did?

Reply:
Let me take a look at it again with fresh eyes. I had to draw out the flow and the relationship between the components (ensurer vs repo controller). I was trying to find a way to have a single component responsible for the creation and initialization of the repo, instead of having that split across 2 (the ensurer that creates the ResticRepository, and the controller that initializes it in object storage and sets the phase to Ready).